### PR TITLE
Relax restrictions on JWTs on write endpoints in JSON API

### PR DESCRIPTION
**Pull request by cocreature**
The previous restriction was both too lax and too strict now that we
have multi pary queries:

1. It allowed a party in `readAs` for command submissions which just
   fails on the ledger side. Changing this is technically breaking but
   only if you used a token that would have been rejected as soon as you
   enabled auth so that seems very resonable to break.

2. It didn’t allow extra parties in `readAs`.

This PR switches to requiring exactly one party in `actAs` while
supporting multiple parties in `readAs`.

changelog_begin

- [JSON API] JWTs on command submissions can now contain extra parties
  in the `readAs` field. `actAs` is still limited to a single party.

changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/EndpointsCompanion.scala [hunk 1]
```diff
@@ -61,13 +61,16 @@ object EndpointsCompanion {
                     Unauthorized("ledgerId missing in access token"))
                   applicationId <- payload.applicationId.toRightDisjunction(
                     Unauthorized("applicationId missing in access token"))
-                  party <- payload.party.toRightDisjunction(
-                    Unauthorized("party missing or not unique in access token"))
+                  actAs <- payload.actAs match {
+                    case p :: Nil => \/-(p)
```
Comment (S11001001) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/EndpointsCompanion.scala:65]:
```suggestion
                    case Seq(p) => \/-(p)
```

Matching on `::` is quite dangerous in Scala.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/EndpointsCompanion.scala [hunk 2]
```diff
@@ -61,13 +61,16 @@ object EndpointsCompanion {
...
+                    case ps => -\/(Unauthorized(s"Expected exactly one party in actAs but got $ps"))
+                  }
                 } yield
                   JwtWritePayload(
                     lar.LedgerId(ledgerId),
                     lar.ApplicationId(applicationId),
-                    lar.Party(party)
+                    lar.Party(actAs),
+                    payload.readAs.map(lar.Party(_))
```
Comment (S11001001) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/EndpointsCompanion.scala:73]:
```suggestion
                    lar.Party.subst(payload.readAs),
```

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/EndpointsCompanion.scala (hunk 1 above)
Comment (cocreature) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/EndpointsCompanion.scala:65]: Why? We are working with `List` here, shouldn’t `::` be fine? Happy to make the change, I’d just like to understand why this is better.

Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/EndpointsCompanion.scala]: As long as the selector is `List` then it is fine. But it is very easy to change the selector's type to `Seq`, and `::` will continue to compile without warning, so the change will almost certainly go unnoticed, especially if it is separate from the type declaration as here. `::` will work for any case where the `Seq` happens to be a `List`, but if any code path uses `Vector` or anything else the `::` will simply not match.

**Review by S11001001 — APPROVED** [review]

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/EndpointsCompanion.scala (hunk 1 above)
Comment (cocreature) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/EndpointsCompanion.scala]: Makes sense, thanks for the explanation!

**Pull request merged**